docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,257 @@
+
+> We'd like to stress that a secure best practice for enclave interfaces is to _avoid_ custom marshaling. But we understand that there may be cases where avoiding it is impractical, so in these sections we will help developers do it safely.
+
+The ecall function implementation below has several problems. We will be correcting this code as we move through the material. The general impetus for custom marshaling in this sample code is to avoid expensive intermediate copying of shared memory introduced by the oeedger8r's generated code. (Data passed to `render_data()` is "flat", that is, the data format is non-referential and thus non-_self_-referencing. These attributes reduce threats to enclaves from untrusted data streams. More on this in the following sections.)
+
+```c++
+typedef struct _blob {
+    void* data;
+    size_t size;
+} blob_t;
+
+// WARNING: Portions of this code are intentionally flawed to demonstrate common pitfalls.
+int ecall_with_user_check0(void* ptr) {
+
+    blob_t* blob;
+
+    if (ptr == nullptr)
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    if (blob->size > 4096)
+        return -1;
+
+    return render_data(blob->data, blob->size);
+}
+```
+
+## Ensuring memory is where it should be
+
+Remember that the code we are focused on is running _inside_ of the enclave. The host caller invoked this function from _outside_ the enclave, passing in a pointer to host memory containing well-formed data (in the non-malicious case). But in the Open Enclave security model, the host caller is untrusted - we must treat untrusted input with caution. 
+
+One important security check is to ensure that memory blocks are on the correct side of the security boundary. It's common for host applications to pass host memory blocks to the enclave, as is the case with this example code.
+
+Let's update the code to ensure the memory block is located _strictly_ outside the secured enclave memory region before operating on it.
+
+```diff
+// WARNING: Portions of this code are intentionally flawed to demonstrate common pitfalls.
+int ecall_with_user_check1(void* ptr) {
+
+    blob_t* blob;
+
+-   if (ptr == nullptr)
+-       return -1;
++   // Ensure passed-in pointer is not null and the memory block is located entirely outside of the enclave.
++   if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
++       return -1;
+
+    blob = (blob_t*)ptr;
+
+    if (blob->size > 4096)
```

Comment:
Significance of comparing the size to 4096 bytes is not obvious

Reply:
I added a comment.
>// blobs to be rendered are limited to 4K